### PR TITLE
Compression handling

### PR DESCRIPTION
**Pull request by howryu**

### CODE REVIEW

File: cpp/FileCompressionType.cpp [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+const FileCompressionType FileCompressionType::NONE =
+  FileCompressionType("", "", true);
+
+bool Snowflake::Client::FileCompressionType::matchMagicNumber(char *header) const
+{
+  if (m_magicNumbers)
+  {
+    for (size_t i = 0; i < m_magicNumbers->size(); i++)
+    {
+      if (!memcmp(header, m_magicNumbers->at(i), m_magicBytes))
```
Comment (smtakeda) [inline cpp/FileCompressionType.cpp]: You may wan to check if data in header is less than m_magicBytes and return false.

File: tests/test_simple_put.cpp [hunk 2]
```diff
@@ -128,17 +137,68 @@ static int teardown(void **unused)
...
 int main(void) {
-  initialize_test(SF_BOOLEAN_FALSE);
   const struct CMUnitTest tests[] = {
-    cmocka_unit_test_setup_teardown(test_simple_put, NULL, teardown),
+    cmocka_unit_test_teardown(test_simple_put_auto_compress, teardown),
+    cmocka_unit_test_teardown(test_simple_put_auto_detect_gzip, teardown),
+    cmocka_unit_test_teardown(test_simple_put_no_compress, teardown),
+    cmocka_unit_test_teardown(test_simple_put_gzip, teardown),
+    cmocka_unit_test_teardown(test_simple_put_auto_detect_bz2, teardown),
+    cmocka_unit_test_teardown(test_simple_put_auto_detect_zst, teardown),
```
Comment (smtakeda) [inline tests/test_simple_put.cpp]: please add the 0 and 1 byte length file test.

**Comment by howryu** [comment]
One of the travis check failed because of `sudo apt get update` failed.

**Comment by smtakeda** [comment]
You can restart the job. if not, let me know.

**Comment by howryu** [comment]
@smtakeda It's hard to say whether `header` will contains null in middle or not since this is just first 4 bytes of any file. Not sure if it is possible that a file's first 4 byte will contain null or not. However, I already preallocate enough space (5 bytes including `\0`). And `memcmp` takes an argument indicating how many bytes to compare, in this case is `m_magicBytes`. Idea is that m_magicBytes will never be bigger than MAX_MAGIC_BYTES macro. So I think we are good here.

**Pull request merged**